distributed/tests/test_client.py

```diff
@@ -8504,3 +8437,41 @@ def test_release_persisted_collection_sync(c):
         # submitting to the scheduler is different to what we are in
         # client.compute
         arr.compute()
+
+
+@pytest.mark.slow()
+@pytest.mark.parametrize("do_wait", [True, False])
+def test_worker_clients_do_not_claim_ownership_of_serialize_futures(c, do_wait):
```

Comment:
I think this test describes the _subtleties_ involved in this change. The cases that raise a `CancelledError` could previously still work, depending on timing. If the futures would unpack on the worker before the client side release reached the scheduler, the futures would still be referenced.
In an async test, this would rather trivially be true since the scheduler also deserializes the future and given that the test client and scheduler run in the same thread, the `get_client` discovery would detect the client even while inside of the scheduler which is the "bug/feature" I had to get rid of here.

https://github.com/dask/distributed/issues/7498 describes this race condition in detail